File: problems/树/222.完全二叉树的节点个数.py
#!/usr/bin/env python3
"""
222. 完全二叉树的节点个数
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def count_nodes(root: Optional[TreeNode]) -> int:
    if not root: return 0
    left_depth = right_depth = 0
    left_node = root
    while left_node:
        left_depth += 1
        left_node = left_node.left
    right_node = root
    while right_node:
        right_depth += 1
        right_node = right_node.right
    if left_depth == right_depth:
        return (1 << left_depth) - 1
    return 1 + count_nodes(root.left) + count_nodes(root.right)

if __name__ == "__main__":
    root1 = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
    print(count_nodes(root1))  # 7
    print(count_nodes(TreeNode(1)))  # 1
    print(count_nodes(None))  # 0

File: problems/树/222.完全二叉树的节点个数.cpp
#include <iostream>
#include <cmath>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int countNodes(TreeNode* root) {
    if (!root) return 0;
    TreeNode *l = root, *r = root;
    int hl = 0, hr = 0;
    while (l) { hl++; l = l->left; }
    while (r) { hr++; r = r->right; }
    if (hl == hr) return (1 << hl) - 1;
    return 1 + countNodes(root->left) + countNodes(root->right);
}
int main() {
    TreeNode* root = new TreeNode(1);
    root->left  = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left  = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->left  = new TreeNode(6);
    root->right->right = new TreeNode(7);
    cout << countNodes(root) << endl;       // 7
    cout << countNodes(new TreeNode(1)) << endl;  // 1
    cout << countNodes(nullptr) << endl;          // 0
    return 0;
}
